### PR TITLE
Replace and deprecate `DataSet` use in class names

### PR DESCRIPTION
**Pull request by deepyaman**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

Deprecate core "DataSet" names and warn about the impending rename to "Dataset".

Once 0.19.0 is released, the deprecated aliases (and even the utils, if so desired) can be removed.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes

### CODE REVIEW

**Comment by merelcht** [comment]
@deepyaman I'm not sure it actually makes sense to the rename on the Kedro framework side. It's a breaking change and we're removing all datasets from here when releasing `0.19.0` anyway.

**Comment by deepyaman** [comment]
> @deepyaman I'm not sure it actually makes sense to the rename on the Kedro framework side. It's a breaking change and we're removing all datasets from here when releasing `0.19.0` anyway.

@merelcht Won't the datasets in `kedro.io.core` stay in core framework, even in 0.19.0? Looking at https://github.com/kedro-org/kedro/issues/2129, are you suggesting that we would keep `DataSet` in the core repo and `Dataset` in plugins?

**Comment by merelcht** [comment]
Ignore me, I should've had my morning coffee before commenting 😂 

We should rename all of them, but it will be a breaking change so maybe it's better to wait until we start work on `0.19.0` to make sure `develop` and `main` don't diverge too much until then.

**Comment by merelcht** [comment]
If we do a deprecation warning what will be users see and how often? It might be annoying if there's nothing they can do about it..

**Comment by deepyaman** [comment]
> If we do a deprecation warning what will be users see and how often? It might be annoying if there's nothing they can do about it..

If this functionality were to be released, they would get a warning if they used `MemoryDataSet` (or `pandas.CSVDataSet` in an updated `kedro-datasets`); but they could change their code to `MemoryDataset` (or `pandas.CSVDataset`) and avoid the warning. So the idea would be, yes, they can do something about it, and they have a grace period until it's removed that their projects wouldn't be broken on a newer version.

File: kedro/utils.py [hunk 1]
```diff
@@ -26,3 +27,64 @@ def load_obj(obj_path: str, default_obj_path: str = "") -> Any:
...
+
+        if alias is not None:
+
+            def new(cls, *args, **kwargs):
+                alias = getattr(cls, "_DeprecatedClassMeta__alias")
+
+                if alias is not None:
+                    warn(
+                        f"{cls.__name__} has been renamed to {alias.__name__}, and the "
+                        f"alias will be removed in the future",
```
Comment (merelcht) [inline kedro/utils.py:49]:
```suggestion
                        f"alias will be removed in the Kedro 0.19.0",
```

File: kedro/utils.py [hunk 2]
```diff
@@ -26,3 +27,64 @@ def load_obj(obj_path: str, default_obj_path: str = "") -> Any:
...
+
+        fixed_bases = []
+
+        for b in bases:
+            alias = getattr(b, "_DeprecatedClassMeta__alias", None)
+
+            if alias is not None:
+                warn(
+                    f"{b.__name__} has been renamed to {alias.__name__}, and the alias "
+                    f"will be removed in the future",
```
Comment (merelcht) [inline kedro/utils.py:67]:
```suggestion
                    f"will be removed in Kedro 0.19.0",
```

File: kedro/utils.py (hunk 1 above)
Comment (deepyaman) [reply kedro/utils.py:49]: @merelcht Do we have another release (or few) before 0.19.0? With this change, we could also warn in 0.19.x at drop in 0.20.0 (or something), if there's any reason to give more time.

If there is no release before 0.19.0, then of course doesn't really help to have this.

Comment (merelcht) [reply kedro/utils.py:49]: Yes we'll definitely have a couple more releases before `0.19.0`.

**Comment by merelcht** [comment]
@deepyaman Yes, let's keep this simple and only do the non-abstract classes. When we're working on `0.19.0` we can properly rename everything here and in `kedro-datasets`.

File: RELEASE.md [hunk 3]
```diff
@@ -162,7 +162,7 @@ Many thanks to the following Kedroids for contributing PRs to this release:
 * Added support for `tf.device` in `TensorFlowModelDataset`.
 * Updated error message for `VersionNotFoundError` to handle insufficient permission issues for cloud storage.
 * Updated Experiment Tracking docs with working examples.
-* Updated MatplotlibWriter Dataset, TextDataset, plotly.PlotlyDataSet and plotly.JSONDataSet docs with working examples.
+* Updated MatplotlibWriter Dataset, TextDataSet, plotly.PlotlyDataSet and plotly.JSONDataSet docs with working examples.
```
Comment (deepyaman) [inline RELEASE.md:165]:
```suggestion
* Updated `MatplotlibWriter`, `text.TextDataSet`, `plotly.PlotlyDataSet` and `plotly.JSONDataSet` docs with working examples.
```

**Comment by stichbury** [comment]
You are truly the patron saint of consistency @deepyaman and I'm happy to approve this change because it seems wholesale and straightforward. However, I'm not able to speak for @merelcht and @idanov in terms of whether it's the right point to merge this change or the implications thereof. But if they're happy with the intent, I'm happy with the execution.

**Review by merelcht — APPROVED** [review]
Thanks for persisting with this @deepyaman ! I'm happy for this to get merged ⭐

**Review by stichbury — APPROVED** [review]
If it's OK with @merelcht it's a 👍 from me too.

File: kedro/utils.py [hunk 4]
```diff
@@ -26,3 +27,65 @@ def load_obj(obj_path: str, default_obj_path: str = "") -> Any:
...
+
+        if alias is not None:
+
+            def new(mcs, *args, **kwargs):
+                alias = getattr(mcs, "_DeprecatedClassMeta__alias")
+
+                if alias is not None:
+                    warn(
+                        f"{mcs.__name__} has been renamed to {alias.__name__}, and the "
+                        f"alias will be removed Kedro 0.19.0",
```
[the remark this answers is not in the document]
Comment (deepyaman) [reply kedro/utils.py:49]: Ah, thanks! I fixed it in a separate place but forgot to add it back in here. :)

**Comment by astrojuanlu** [comment]
RTD failing with:

```
sphinx.errors.SphinxWarning: /home/docs/checkouts/readthedocs.org/user_builds/kedro/checkouts/2500/kedro/extras/datasets/api/api_dataset.py:docstring of kedro.io.core.AbstractDataSet.exists:1:py:exc reference target not found: DatasetError
```

**Pull request merged**